routemaster/state_machine/api.py

```diff
@@ -113,186 +89,147 @@ def create_label(app: App, label: Label, metadata: Metadata) -> Metadata:
         if deleted:
             raise DeletedLabel(label)
 
-        new_metadata = dict_merge(existing_metadata, update)
-
-        conn.execute(labels.update().where(label_filter).values(
-            metadata=new_metadata,
-        ))
-
-        _move_label_for_metadata_change(
+        needs_gate_evaluation = needs_gate_evaluation_for_metadata_change(
             state_machine,
             label,
             update,
-            new_metadata,
             conn,
         )
 
-        return new_metadata
-
-
-def _move_label_for_metadata_change(
-    state_machine: StateMachine,
-    label: Label,
-    update: Metadata,
-    metadata: Metadata,
-    conn,
-) -> None:
-    history_entry = conn.execute(
-        select([history]).where(and_(
-            history.c.label_name == label.name,
-            history.c.label_state_machine == label.state_machine,
-        )).order_by(
-            history.c.created.desc(),
-        ).limit(1)
-    ).fetchone()
-
-    current_state = state_machine.get_state(history_entry.new_state)
-    if isinstance(current_state, Action):
-        # Label is in an Action state so there's no trigger to resolve.
-        return
-
-    if not any(
-        trigger.should_trigger_for_update(update)
-        for trigger in current_state.metadata_triggers
-    ):
-        return
-
-    feeds = feeds_for_state_machine(state_machine)
-    exit_condition_context = Context(
-        label.name,
-        metadata,
-        _utcnow(),
-        feeds,
-        current_state.exit_condition.accessed_variables(),
-    )
-    can_exit = current_state.exit_condition.run(exit_condition_context)
-
-    if not can_exit:
-        return
-
-    destination = _choose_destination(
-        state_machine,
-        current_state,
-        exit_condition_context,
-    )
-
-    conn.execute(history.insert().values(
-        label_name=label.name,
-        label_state_machine=state_machine.name,
-        old_state=current_state.name,
-        new_state=destination.name,
-    ))
-
-
-def _choose_destination(
-    state_machine: StateMachine,
-    current_state: State,
-    context: Context,
-) -> State:
-    next_state_name = current_state.next_states.next_state_for_label(context)
-    return state_machine.get_state(next_state_name)
+        new_metadata = dict_merge(existing_metadata, update)
 
+        conn.execute(labels.update().where(and_(
+            labels.c.name == label.name,
+            labels.c.state_machine == label.state_machine,
+        )).values(
+            metadata=new_metadata,
+            metadata_triggers_processed=not needs_gate_evaluation,
+        ))
 
-def delete_label(app: App, label: Label) -> None:
+    # Outside transaction
+    if needs_gate_evaluation:
+        try:
+            with app.db.begin() as conn:
+                lock_label(label, conn)
+                current_state = get_current_state(label, state_machine, conn)
+                if not isinstance(current_state, Gate):  # pragma: no branch
```

Comment:
While I agree that this is here mostly for the benefit of `mypy`, I'd feel happier if it was outside the `try..except` which silences errors.

Reply:
Scratch that; of course it can't be moved. Should we consider using `cast` rather than this check? Alternatively, could we catch & re-raise this error separately?

I'm wary that we have errors being completely silenced (not even logged).

Reply:
Hmmmm. Not sure what's best here. I'll go with the loud error for now. a8ef1917d0a496e46fa15fc7036eb13a0bf85bcc